### PR TITLE
[bitfinex] tickers

### PR DESCRIPTION
**Pull request by andre77**
fixed the issue with different types of tickers
there are currently 2:
 - trading pairs starting with t, like tBTCUSD
 - funding currencies starting with f, like fLEO
see https://docs.bitfinex.com/reference#rest-public-tickers

### CODE REVIEW

File: xchange-bitfinex/src/main/java/org/knowm/xchange/bitfinex/service/BitfinexMarketDataServiceRaw.java [hunk 1]
```diff
@@ -92,15 +93,19 @@ public BitfinexLendDepth getBitfinexLendBook(String currency, int limitBids, int
 
   public org.knowm.xchange.bitfinex.v2.dto.marketdata.BitfinexTicker[] getBitfinexTickers(
       Collection<CurrencyPair> currencyPairs) throws IOException {
-    return bitfinexV2.getTickers(BitfinexAdapters.adaptCurrencyPairsToTickersParam(currencyPairs));
+    List<ArrayNode> tickers =
+        bitfinexV2.getTickers(BitfinexAdapters.adaptCurrencyPairsToTickersParam(currencyPairs));
```
Comment (walec51) [inline xchange-bitfinex/src/main/java/org/knowm/xchange/bitfinex/service/BitfinexMarketDataServiceRaw.java]: this PR adds support for "fUSD" bitfinex symbols
however I do not see any way that `BitfinexAdapters.adaptCurrencyPairsToTickersParam` would ever produce and "fUSD" symbol for any given CurrencyPair

em I missing something?

ps. besides that its hard to view "fUSD" as a currency pair, what does the opening price of a fUSD symbol mean?

Comment (andre77) [reply xchange-bitfinex/src/main/java/org/knowm/xchange/bitfinex/service/BitfinexMarketDataServiceRaw.java]: @walec51 
its not just a currency, but funding currency
basically just have a look at the api spec here https://docs.bitfinex.com/reference#rest-public-tickers
here is the request, and how the data looks like: https://api-pub.bitfinex.com/v2/tickers?symbols=ALL

again, there is an array of data, which can have different types
 - trading pairs starting with t, like tBTCUSD (array with 11 fields)
 - funding currencies starting with f, like fLEO (array with 17 fields)

Comment (walec51) [reply xchange-bitfinex/src/main/java/org/knowm/xchange/bitfinex/service/BitfinexMarketDataServiceRaw.java]: I did read their documentation, however I still do not understand how a fLEO will working in the context of

```
  public static Ticker adaptTicker(
      org.knowm.xchange.bitfinex.v2.dto.marketdata.BitfinexTicker bitfinexTicker) {

    BigDecimal last = bitfinexTicker.getLastPrice();
    BigDecimal bid = bitfinexTicker.getBid();
    BigDecimal bidSize = bitfinexTicker.getBidSize();
    BigDecimal ask = bitfinexTicker.getAsk();
    BigDecimal askSize = bitfinexTicker.getAskSize();
    BigDecimal high = bitfinexTicker.getHigh();
    BigDecimal low = bitfinexTicker.getLow();
    BigDecimal volume = bitfinexTicker.getVolume();

    CurrencyPair currencyPair =
        CurrencyPairDeserializer.getCurrencyPairFromString(bitfinexTicker.getSymbol().substring(1));

    return new Ticker.Builder()
        .currencyPair(currencyPair)
        .last(last)
        .bid(bid)
        .ask(ask)
        .high(high)
        .low(low)
        .volume(volume)
        .bidSize(bidSize)
        .askSize(askSize)
        .build();
  }
```

what will be the resulting currency pair for this ticket in the case of a bitfinex symbol `fLEO`?
what will be the `base` and what will be he `counter`?

could you add a unit test that answers my question?

Comment (andre77) [reply xchange-bitfinex/src/main/java/org/knowm/xchange/bitfinex/service/BitfinexMarketDataServiceRaw.java]: well, this is a good question, and actually not part of this PR
the PR fixes the following error:

this method https://api-pub.bitfinex.com/v2/tickers?symbols=ALL returns unfortunatelly an array with different types:
 - trading pairs

`[
    SYMBOL,
    BID, 
    BID_SIZE, 
    ASK, 
    ASK_SIZE, 
    DAILY_CHANGE, 
    DAILY_CHANGE_RELATIVE, 
    LAST_PRICE, 
    VOLUME, 
    HIGH, 
    LOW
  ]`
for example:
`[
"tBTCUSD",
7326.3,
34.31325066,
7326.4,
14.556629819999998,
-46.28079101,
-0.0063,
7326.31920899,
3956.40350853,
7422,
7297.3
]`
 - funding currencies

`[
    SYMBOL,
    FRR, 
    BID, 
    BID_PERIOD,
    BID_SIZE, 
    ASK, 
    ASK_PERIOD,
    ASK_SIZE,
    DAILY_CHANGE,
    DAILY_CHANGE_RELATIVE, 
    LAST_PRICE,
    VOLUME,
    HIGH, 
    LOW,
    _PLACEHOLDER,
    _PLACEHOLDER,
    FRR_AMOUNT_AVAILABLE
  ]`
for example:
`[
"fLEO",
1.0958904109589042e-8,
0,
0,
0,
1e-8,
2,
2656344.95799559,
0,
0,
1e-8,
400,
1e-8,
1e-8,
null,
null,
2594257.74114297
]`

this is the class [BitfinexTicker.java](https://github.com/knowm/XChange/blob/develop/xchange-bitfinex/src/main/java/org/knowm/xchange/bitfinex/v2/dto/marketdata/BitfinexTicker.java) where the data is parsed , and if you have a closer look, you will see for example that the field VOLUME is in the first array at 9th position and in the second at 12th.
after this PR additional fields for funding currencies were added (like frr, bid, etc), which has broken the code https://github.com/knowm/XChange/pull/3322

i just fixed that point.
i did not touch the method adaptTicker

Comment (walec51) [reply xchange-bitfinex/src/main/java/org/knowm/xchange/bitfinex/service/BitfinexMarketDataServiceRaw.java]: `MarketDataService#getTickers` implementation for Birfinex uses the DTOs and Raw methods that you modified and passes their results to `adaptTicker` so saying that my question is not part of this PR is like saying that testing code that uses the code you modified is not relevant

if Bitfinexes "founding currency ticker" has no mapping to the generic Ticker that makes sens (we have a currency pair their as an attribute - not a currency) then modify  `BitfinexMarketDataService#getTickers` to ignore them

Comment (andre77) [reply xchange-bitfinex/src/main/java/org/knowm/xchange/bitfinex/service/BitfinexMarketDataServiceRaw.java]: this [PR 3322](https://github.com/knowm/XChange/pull/3322) has broken the code, just try to fetch tickers and have a look at the volume or other fields...
my PR fixes it, as simple as that
feel free to close it, i really don't mind

Comment (walec51) [reply xchange-bitfinex/src/main/java/org/knowm/xchange/bitfinex/service/BitfinexMarketDataServiceRaw.java]: yeah, I merged https://github.com/knowm/XChange/pull/3322 and it was a mistake

but then again we are all used to merging untested and narrow minded here

and I guess we wont start enforcing better practices starting from this pull request

**Pull request merged**